wandb/dummy.py

```diff
@@ -0,0 +1,201 @@
+        pass
+
+    def __irshift__(self, other):
+        pass
+
+    def __iand__(self, other):
+        pass
+
+    def __ixor__(self, other):
+        pass
+
+    def __ior__(self, other):
+        pass
+
+    def __neg__(self):
+        return Dummy()
+
+    def __pos__(self):
+        return Dummy()
+
+    def __abs__(self):
+        return Dummy()
+
+    def __invert__(self):
+        return Dummy()
+
+    def __complex__(self):
+        return 1 + 0j
+
+    def __int__(self):
+        return 1
+
+    def __long__(self):
+        return 1
+
+    def __float__(self):
+        return 1.0
+
+    def __oct__(self):
+        return oct(1)
+
+    def __hex__(self):
+        return hex(1)
+
+    def __lt__(self, other):
+        return True
+
+    def __le__(self, other):
+        return True
+
+    def __eq__(self, other):
+        return True
+
+    def __ne__(self, other):
+        return True
+
+    def __gt__(self, other):
+        return True
+
+    def __ge__(self, other):
+        return True
+
+    def __getattr__(self, attr):
+        try:
+            return object.__getattribute__(self, attr)
+        except AttributeError:
+            dummy = Dummy()
+            setattr(self, attr, dummy)
+            return dummy
+
+    def __getitem__(self, key):
+        if key in self.___dict:
+            return self.___dict[key]
+        dummy = Dummy()
+        self.___dict[key] = dummy
+        return dummy
+
+    def __setitem__(self, key, value):
+        self.___dict[key] = value
+
+    def __call__(self, *args, **kwargs):
+        return Dummy()
+
+    def __len__(self):
+        return 1
+
+    def __str__(self):
+        return ""
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        return exc_type is None
+
+    def __repr__(self):
+        return ""
+
+
+class DummyModule(Dummy, ModuleType):
+    pass
+
+
+def disable(globals_=None):
+    import wandb
+
+    wandb_module = wandb._wandb_module
+    for m in sys.modules:
+        if m == wandb_module or (
+            m.startswith(wandb_module + ".")
+            and not m.startswith(wandb_module + ".dummy")
+        ):
+            sys.modules[m] = DummyModule()
+    if globals_:
+        for k, v in globals_.items():
+            m = getattr(
+                v, "__module__", v.__name__ if isinstance(v, ModuleType) else ""
+            )
+            if m == wandb_module or (
+                m.startswith(wandb_module + ".")
+                and not m.startswith(wandb_module + ".dummy")
+            ):
+                globals_[k] = Dummy()
+
+    class ImportHook:
+        def find_module(self, fullname, path=None):
+            if fullname.startswith(wandb_module + "."):
+                return self
+
+        def load_module(self, fullname):
+            module = DummyModule()
+            sys.modules[fullname] = module
+            return module
+
+    sys.meta_path.insert(0, ImportHook())
```

Comment:
Does this code looks very magical.  Can you add a doc string that describes it's purpose, where it is used, and what the ImportHook is for?

Reply:
comment added